### PR TITLE
feat: gsheets schema UI

### LINKED ISSUE #25647: [Feature]: Fetch spreadsheet, sheets and sheet data and store it in redux
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

1. As soon as the gsheet is authorised and we come back to appsmith page, we should make following three calls:
    1. `/trigger` call to fetch list of spreadsheets
        1. This call needs to be made as soon as we come to appsmith page
    2. `/trigger` call to fetch list of sheets
        1. This call needs to be made when first spreadsheet is populated by default in the spreadsheet dropdown
    3. `/trigger` call to fetch the sheet data
        1. This call needs to be made when first sheet from first spreadsheet is populated by default in the sheet dropdown
2. We need to store all above fetched data in redux so that it can be used to display data
3. When sheet is not authorised, [TBD](https://www.figma.com/file/kbU9xPv44neCfv1FFo9Ndu/User-Activation?node-id=1274%3A63719&mode=dev)

### Why should this be worked on?

When gsheet datasource is authorised, as a frontend engineer I should fetch be able to spreadsheets, sheets and sheet data to store in redux state, so that eventually we can use it display the data in correct format

### LINKED ISSUE #25648: [Feature]: Display all data fetched in the UI
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

1. Once the required data is fetched for spreadsheets, sheets and sheet data, Create a component to display this data on review page.
    1. Add dropdown element for spreadsheet, and populate spreadsheets data in it, by default populate the dropdown with first spreadsheet from list
    2. Add dropdown for sheets, and populate sheets data for a given spreadsheet. By default populate it with first sheet.
    3. Add a table component and pass sheet data to it. Display sheet data in this table

### Why should this be worked on?

When gsheets datasource is authorised, I should be able to see table with data from spreadsheet, so that I can easily connect that data to widgets

### LINKED ISSUE #25649: [Feature]: Add CTA for Create list and detail and loading states
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

1. Add CTA for Create a list and detail on the datasource review page
2. Show loading state for table and dropdown components

### Why should this be worked on?

When I authorise the datasource, I should see CTA for Creating list and detail so that I can use this CTA to connect the data to widgets

### LINKED ISSUE #25650: [Feature]: Integrate Create list and detail with CRUD APIs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

1. Integrate Create list and detail functionality with `/crud-page` API
    1. If current page does not have widgets or queries with names `DeleteQuery`, `SelectQuery`, `UpdateQuery`, `InsertQuery`pass pageId to the API
    2. If current page has widgets/queries, [TBD](https://www.figma.com/file/kbU9xPv44neCfv1FFo9Ndu/User-Activation?node-id=1274%3A63719&mode=dev)

### Why should this be worked on?

As frontend engineer, I should be able to connect Create list and detail to CRUD generation API, so that I can provide the necessary value of this CTA to users

### LINKED ISSUE #25834: [Bug]: All spreadsheets are shown in gsheets schema dropdown, even when selected sheets option is used
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When I create gsheet datasource with selected sheets option, I still see all spreadsheets in the spreadsheet dropdown of gsheet schema. But when I refresh the page, only selected sheets are shown in dropdown as expected

### Steps To Reproduce

1. Create gsheet datasource with selected sheets option
2. Check the spreadsheet dropdown in schema UI

### Public Sample App

_No response_

### Environment

Deploy Preview

### Issue video log

_No response_

### Version

Cloud/ Self-Hosted - 1.8.4

### LINKED ISSUE #26025: [Bug]: Handle empty data from gsheet on gsheet schema
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Loading state doesn't change even if the data trigger is complete and its because of the empty data on the sheet or the header starts with index other than 1.

### Steps To Reproduce

1. Authorize the Google sheet datasource
2. Select the spreadsheet and sheet with empty data
3. Observe the Loading state remains.

### Public Sample App

_No response_

### Environment

Deploy Preview

### Issue video log

_No response_

### Version

Self Hosted

### LINKED ISSUE #26034: [Bug]: Selected gsheet issue in google sheet schema feature
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

On selected Gsheet the user is gained permission to sheet which he has not been given permission to.

### Steps To Reproduce

1. Create gsheet with selected sheets option
2. Observe the schema

### Public Sample App

_No response_

### Environment

Deploy Preview

### Issue video log

https://nwshheskua.vmaker.com/record/i5xXBKpkMaL1nhrl

### Version

Cloud/ Self-Hosted - 1.8.4

### PR DESCRIPTION
**Pull request by sneha122**
## Description
- Added a Gsheet Schema to show a preview of sheet data if the gsheet datasource is authorized successfully. User can change the spreadsheet and specific sheet to check the preview.
- Added feature behind feature flag. Hence, Gsheet Schema is shown to users having feature flag `ab_gsheet_schema_enabled` -> `true`.
- Added `create a list and detail` functionality to generate crud for a gsheet selected `spreadsheet name` and `sheet name`.
- Added analytic events for the gsheet schema preview

#### PR fixes following issue(s)
Fixes #25647
Fixes #25648 
Fixes #25649 
Fixes #25650 
Fixes #25834 
Fixes #26025
Fixes #26034

#### Media
Screenshot 2023-07-28 at 1 46 54 AM

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
- [ ] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+      !isFetchingSheetsList
+    ) {
+      setSelectedSheet(sheetsList[0]);
+    }
+  }, [selectedSheet, sheetsList, isFetchingSheetsList]);
+
+  // Set current sheet data
+  useEffect(() => {
+    if (sheetData?.length > 0) {
+      // Getting the top 12 rows as for experimentation we need to keep this number fixed for preview
```
Comment (Irongade) [inline app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx:195]: Make this a constant. something like MAX_SHEET_LENGTH

File: app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    sheetValue: string | undefined,
+    sheetObj: DropdownOption | undefined,
+  ) => {
+    if (sheetValue && sheetObj) {
+      setSelectedSheet(sheetObj);
+    }
+  };
+
+  const isError = selectedDatasourceIsInvalid || failedFetchingSheetsList;
+  const isLoading =
```
Comment (Irongade) [inline app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx:219]: Why are you centralizing all the loading states? Why not have them separately for each components.

Comment (sneha122) [reply app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx:219]: @Irongade Dropdowns are using their own states to show loading state, This aggregated loading state is only for table and CTA. Please correct me if I am wrong @AmanAgarwal041

File: app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx (hunk 1 above)
Comment (sneha122) [reply app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx:197]: Made changes. Please check

File: app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx (hunk 2 above)
Comment (AmanAgarwal041) [reply app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx:225]: @sneha122 you have explained it correctly.

Comment (Irongade) [reply app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx:225]: Okay, sounds good.

File: app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    ) {
+      setSelectedSheet(sheetsList[0]);
+    }
+  }, [selectedSheet, sheetsList, isFetchingSheetsList]);
+
+  // Set current sheet data
+  useEffect(() => {
+    if (sheetData?.length > 0) {
+      // Getting the top 12 rows as for experimentation we need to keep this number fixed for preview
+      setCurrentSheetData(sheetData.slice(0, 12));
```
Comment (Irongade) [inline app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx]: You have not made this change yet.

Comment (AmanAgarwal041) [reply app/client/src/pages/Editor/SaaSEditor/GoogleSheetSchema.tsx]: done

**Review by Irongade — APPROVED** [review]

**Review by Irongade — APPROVED** [review]

File: app/client/src/ce/entities/FeatureFlag.ts [hunk 4]
```diff
@@ -12,6 +12,8 @@ export const FEATURE_FLAG = {
     "release_embed_hide_share_settings_enabled",
   ab_ds_schema_enabled: "ab_ds_schema_enabled",
   ab_ds_binding_enabled: "ab_ds_binding_enabled",
+  ab_gsheet_schema_enabled: "ab_gsheet_schema_enabled",
+  release_scim_provisioning_enabled: "release_scim_provisioning_enabled",
```
Comment (hetunandu) [inline app/client/src/ce/entities/FeatureFlag.ts:16]: `release_scim_provisioning_enabled` is removed and no longer needed

Comment (AmanAgarwal041) [reply app/client/src/ce/entities/FeatureFlag.ts]: Removed

**Review by AmanAgarwal041 — APPROVED** [review]

**Review by Irongade — APPROVED** [review]

**Pull request merged**